joinservice/internal/certcache/certcache.go

```diff
@@ -0,0 +1,165 @@
+	case variant.AzureSEVSNP{}:
+		c.log.Debugf("Creating azure SEV-SNP certificate chain cache")
+		ask, ark, err := c.createCertChainCache(ctx, abi.VcekReportSigner)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create azure SEV-SNP certificate chain cache: %w", err)
+		}
+		return &CachedCerts{
+			ask: ask,
+			ark: ark,
+		}, nil
+	// TODO(derpsteb): Add AWS
+	default:
+		c.log.Debugf("No certificate chain caching possible for attestation variant %s", *c.attVariant)
+		return nil, nil
+	}
+}
+
+// CachedCerts contains the cached certificates.
+type CachedCerts struct {
+	ask *x509.Certificate
+	ark *x509.Certificate
+}
+
+// SevSnpCerts returns the cached SEV-SNP ASK and ARK certificates.
+func (c *CachedCerts) SevSnpCerts() (ask, ark *x509.Certificate) {
+	return c.ask, c.ark
+}
+
+// createCertChainCache creates a certificate chain cache configmap with the ASK and ARK
+// retrieved from the KDS and returns ASK and ARK.
+// If the configmap already exists, nothing is done and the existing ASK and ARK are returned.
+func (c *Client) createCertChainCache(ctx context.Context, signingType abi.ReportSigner) (ask, ark *x509.Certificate, err error) {
+	c.log.Debugf("Creating certificate chain cache")
+	ask, ark, err = c.getCertChainCache(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get certificate chain cache: %w", err)
+	}
+	if ask != nil && ark != nil {
+		c.log.Debugf("Certificate chain cache already exists")
+		return ask, ark, nil
+	}
+
+	c.log.Debugf("Retrieving certificate chain from KDS")
+	ask, ark, err = c.kdsClient.CertChain(signingType)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to retrieve certificate chain from KDS: %w", err)
+	}
+
+	askWriter := &strings.Builder{}
+	if err := pem.Encode(askWriter, &pem.Block{Type: "CERTIFICATE", Bytes: ask.Raw}); err != nil {
+		return nil, nil, fmt.Errorf("failed to encode ask: %w", err)
+	}
+	arkWriter := &strings.Builder{}
+	if err := pem.Encode(arkWriter, &pem.Block{Type: "CERTIFICATE", Bytes: ark.Raw}); err != nil {
+		return nil, nil, fmt.Errorf("failed to encode ark: %w", err)
+	}
+
+	c.log.Debugf("Creating certificate chain cache configmap")
+	if err := c.kubeClient.CreateConfigMap(ctx, constants.SevSnpCertCacheConfigMapName, map[string]string{
+		constants.CertCacheAskKey: askWriter.String(),
+		constants.CertCacheArkKey: arkWriter.String(),
+	}); err != nil {
+		return nil, nil, fmt.Errorf("failed to create certificate chain cache configmap: %w", err)
+	}
+
+	return ask, ark, nil
+}
+
+// getCertChainCache returns the cached ASK and ARK certificate.
+func (c *Client) getCertChainCache(ctx context.Context) (ask, ark *x509.Certificate, err error) {
+	c.log.Debugf("Retrieving certificate chain from cache")
+	askRaw, err := c.kubeClient.GetConfigMapData(ctx, constants.SevSnpCertCacheConfigMapName, constants.CertCacheAskKey)
+	if err != nil {
+		if _, ok := err.(kubernetes.ConfigMapNotExistError); ok {
+			return nil, nil, nil
+		}
```

Comment:
Instead of returning a `nil` error, and then checking on the return values in the caller, this function should let the caller handle checking for specific errors and then decide on what actions to take.